### ISSUE TITLE
Unintuitive recompile in the Latent ODE example

### ISSUE THREAD

**Issue by nicholasjng** [patrick-kidger/diffrax#111]
So I am looking to start a small side project on neural ODEs, and finally got time to properly go over the examples here.

I noticed that in the latent ODE example, the 39th step takes orders of magnitude longer than the others (much like the first overall step, but that's obviously the one that carries the JIT compilation overhead). On my local machine, it is about 21 secs (step 0) vs. 48 secs (step 39).

After importing the JAX config, and setting `jax_log_compiles` to `True`, I got confirmation that JAX is indeed recompiling `make_step` there. Curiously, it never recompiles after that again. The first compilation of `make_step` takes 17.6 seconds, while the second one takes 44.1 seconds.

After looking at the example a little bit, I think I recognized the significance of 39 here, as that is the integer result of the dataset size (10000) divided by the batch size (256). What I do not yet understand is why there is always a cache hit after the first recompile. The example uses `filter_jit` from Equinox, which I am not at all familiar with (although I believe that it is `jit` together with a pytree partition op). Could that have something to do with it?

So in short, two questions:
1. Why does it recompile (presumably after the first dataloader exhaust) ?
2. Why does the recompile take 2.5 times as long ?

**Comment by patrick-kidger** [comment]
So the recompilation will be because at this point the batch size changes. It will be slightly smaller as we're at the end of the dataset.

If you're adapting this example for your own use case then it'll be pretty easy to adjust the dataloader to just drop the last batch instead.

If you're pursuing a larger or more serious project then I'd recommend using the PyTorch dataloaders. They have very little specialisation on the rest of PyTorch so they're a good choice to use even in a JAX project.

As for why the compilation takes longer the second time around: no idea. I agree that sounds a little mysterious. If you're  curious enough to dig into it then I'd recommend a tool like [py-spy](https://github.com/benfred/py-spy) to create flamegraphs of the compilation process.

**Comment by nicholasjng** [comment]
Correct! Rounding up to the next multiple of 256 makes the second recompile go away. Thanks for the hint, should've seen that myself (although I guess, that is the work of Equinox abstracting away the static arguments machinery from the JIT).

I'll try and see if I can come up with something about the recompilation if I have time, but I will definitely go with a dedicated dataloader instead. Thanks for the recommendation on Pytorch - do you happen to know how they compare to TF's `dataset`s ?

Anyway, I will close this to tidy up the issue tracker, if you don't mind. :)

**Comment by patrick-kidger** [comment]
Glad to have helped!

From what I hear most projects use either TF datasets or PyTorch dataloaders, but I don't know how they compare.
